temporal-sdk/src/main/java/io/temporal/internal/activity/ActivityTaskExecutors.java

```diff
@@ -128,6 +128,12 @@ public ActivityTaskHandler.Result execute(ActivityInfoInternal info, Scope metri
             metricsScope,
             local,
             dataConverterWithActivityContext);
+      } finally {
+        if (!context.isDoNotCompleteOnReturn()) {
+          // if the activity is not completed, we need to cancel the heartbeat
+          // to avoid sending it after the activity is completed
+          context.cancelOutstandingHeartbeat();
```

Comment:
Can you clarify this a bit? So except for async-activity-completion, this is going to prevent a pending heartbeat from being recorded? Is there somewhere else in Java where the last heartbeat of a failing activity is guaranteed to be sent?

Reply:
> Is there somewhere else in Java where the last heartbeat of a failing activity is guaranteed to be sent?

Yes, we send the last heartbeat value as part of the activity failure.